src/oscar/apps/search/views.py

```diff
@@ -1,73 +1,121 @@
+class BaseSearchView:
+    template_name = "oscar/search/results.html"
+    context_object_name = "results"
 
+    def dispatch(self, request, *args, **kwargs):
         # Raise a signal for other apps to hook into for analytics
-        self.search_signal.send(
+        user_search.send(
             sender=self,
             session=self.request.session,
             user=self.request.user,
-            query=self.query,
+            query=self.request.GET.get("q"),
         )
 
-        return response
-
-    # Override this method to add the spelling suggestion to the context and to
-    # convert Haystack's default facet data into a more useful structure so we
-    # have to do less work in the template.
-    def extra_context(self):
-        extra = super().extra_context()
-
-        # Show suggestion no matter what.  Haystack 2.1 only shows a suggestion
-        # if there are some results, which seems a bit weird to me.
-        if self.results.query.backend.include_spelling:
-            # Note, this triggers an extra call to the search backend
-            suggestion = self.form.get_suggestion()
-            if suggestion != self.query:
-                extra["suggestion"] = suggestion
-
-        # Convert facet data into a more useful data structure
-        if "fields" in extra["facets"]:
-            munger = FacetMunger(
-                self.request.get_full_path(),
-                self.form.selected_multi_facets,
-                self.results.facet_counts(),
-            )
-            extra["facet_data"] = munger.facet_data()
-            has_facets = any(
-                [len(data["results"]) for data in extra["facet_data"].values()]
-            )
-            extra["has_facets"] = has_facets
-
-        # Pass list of selected facets so they can be included in the sorting
-        # form.
-        extra["selected_facets"] = self.request.GET.getlist("selected_facets")
-
-        return extra
-
-    def get_results(self):
-        # We're only interested in products (there might be other content types
-        # in the Solr index).
-        return super().get_results().models(Product)
+        return super().dispatch(request, *args, **kwargs)
+
+
+class FacetedSearchView(BaseSearchView, BaseFacetedSearchView):
+    form_class = SearchForm
+
+
+class BaseCatalogueView:
+    context_object_name = "products"
+    template_name = "oscar/catalogue/browse.html"
+    enforce_paths = True
+
+    def get(self, request, *args, **kwargs):
+        try:
+            return super().get(request, *args, **kwargs)
+        except Http404:
+            # Redirect to page one.
+            messages.error(request, _("The given page number was invalid."))
+            return redirect("catalogue:index")
+
+    def get_context_data(self, *args, **kwargs):
+        ctx = super().get_context_data(*args, **kwargs)
+        ctx["summary"] = _("All products")
+        return ctx
+
+
+class CatalogueView(BaseCatalogueView, BaseFacetedSearchView):
```

Comment:
Why not move this view to apps/catalogue/views and import BaseFacetedSearchView over there.